A86-xds-http-connect.md

```diff
@@ -0,0 +1,108 @@
+
+## Proposal
+
+The xDS configuration for HTTP CONNECT is done via the
+[`Http11ProxyUpstreamTransport`](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/extensions/transport_sockets/http_11_proxy/v3/upstream_http_11_connect.proto#L36)
+transport socket wrapper.  This wrapper is sent in the CDS [`transport_socket`
+field](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/config/cluster/v3/cluster.proto#L1099),
+as a wrapper around whatever transport socket is actually used.
+
+Note that the only transport socket implementation that gRPC currently
+supports is `UpstreamTlsContext`, which is used to configure TLS
+security, as per [gRFC A29].  We honor the xDS TLS configuration only if
+the application uses `XdsCredentials` when creating the gRPC channel.
+With this design, we will now honor the `Http11ProxyUpstreamTransport`
+transport socket wrapper regardless of whether `XdsCredentials` is used.
+However, we will treat the underlying transport socket wrapped by the
+`Http11ProxyUpstreamTransport` wrapper exactly the way that we would
+previously have treated the transport socket specified directly in the CDS
+resource: if the underlying transport socket is `UpstreamTlsContext`, then
+we will honor it only if `XdsCredentials` is used, and if it is any other
+type, we will NACK the CDS resource.
+
+The `Http11ProxyUpstreamTransport` transport socket
+wrapper looks for the proxy address in the EDS metadata,
+both at the individual endpoint level (in the [`LbEndpoint.metadata`
+field](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/config/endpoint/v3/endpoint_components.proto#L122))
+and at the locality level (in the [`LocalityLbEndpoints.metadata`
+field](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/config/endpoint/v3/endpoint_components.proto#L165)).
+
+To support this, we will reuse the metadata registry mechanism described
+for CDS in [gRFC A83] for the EDS metadata fields.  The parsed form
+of the EDS resource will include a parsed metadata map at both the
+individual endpoint and locality levels, using the same data type
+previously introduced for CDS metadata.
+
+We will support the [`envoy.config.core.v3.Address`
+type](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/config/core/v3/address.proto#L175)
+as a registered metadata type.  When validating this type, the
+`socket_address` field must be set.  Inside of that field, the `address`
+field must be set to an IPv4 or IPv6 address, and the `port_value` field
+must be set.
+
+The proxy address for a given endpoint will be set by looking for a
+metadata entry of type `envoy.config.core.v3.Address` under the key
+`envoy.http11_proxy_transport_socket.proxy_address`.  If that key is not
+present, or if the key is present but the value has a different type,
```

Comment:
> or if the key is present but the value has a different type

Is that really possible?

The previous paragraph talks about ensuring that the `Address` type contains a `SocketAddress` and that the `address` field contains a V4 or V6 address. If we do that validation and NACK if the validation fails, when the map is looked up, we would either find an entry with the desired type or will not find one.

Am I missing something?

Reply:
In the xDS protos, it's possible to have an entry in the `metadata.typed_filter_metadata` map with the key `envoy.http11_proxy_transport_socket.proxy_address` and a value of a type other than `envoy.config.core.v3.Address`.

If the type that is present in the proto is not one of the registered metadata types that we support, then it will be ignored at resource validation time (i.e., the entry will simply not exist in the map in the parsed EDS resource).  In this case, when we actually go to do the map lookup, we won't find any matching entry, so the issue of type mismatch can't arise.

However, if the type that is present in the proto *is* a registered metadata type that we support (e.g., it could be the `extensions.filters.http.gcp_authn.v3.Audience` type described in gRFC A83), then the map in the parsed EDS resource will contain an entry with the wrong value type for this key.  In that case, we need to check the value type and ignore the entry.